### PR TITLE
libteec: Move OP-TEE defined fields into an imp struct

### PR DESCRIPTION
**Pull request by JLLinaro**
GlobalPlatform TEE Client API Specification v1.0 specifies that the structs TEEC_Context, TEEC_Session, TEEC_SharedMemory, and TEEC_Operation shall have a user defined struct named imp. In OP-TEE the struct is not there and instead the user defined fields are declared directly in the top structs.
This commit introduces the imp struct to better support using different implementations. The imp fields now represent the implementation defined parts of the structs that was previously declared directly in the top struct. All previously available parameters are preserved in the imp struct.

Link: https://github.com/OP-TEE/optee_client/issues/348
Reported-by: Tom Hebb <tommyhebb@gmail.com>

### CODE REVIEW

**Comment by JLLinaro** [comment]
The automatic tests fail but that has been solved in the following pull request https://github.com/OP-TEE/optee_test/pull/684

**Comment by jforissier** [comment]
This LGTM, but as mentioned in https://github.com/OP-TEE/optee_client/issues/348, it changes the layout of `TEEC_SharedMemory` and therefore is an ABI breakage. So the `libteec` version should be bumped from 1.0.0 to 2.0.0.

**Comment by JLLinaro** [comment]
@jenswi-linaro version is now updated to 2.0.0

**Comment by jforissier** [comment]
We need `MAJOR_VERSION := 2` in `libteec/Makefile`, too.

File: public/tee_client_api.h [hunk 1]
```diff
@@ -293,16 +295,18 @@ typedef struct {
...
+	struct {
+		int id;
+		size_t alloced_size;
+		void *shadow_buffer;
+		int registered_fd;
+		union {
+			bool dummy;
+			uint8_t flags;
+		} internal;
+    } imp;
```
Comment (etienne-lms) [inline public/tee_client_api.h:309]: fix indentation

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit 
"libteec: Move OP-TEE defined fields into an imp struct" 
(with indentation issue fixed)

**Comment by jforissier** [comment]
Please squash the second commit into the first one (and keep subject "libteec: Move OP-TEE defined fields into an imp struct").
With that:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jforissier** [comment]
There is a conflict, would you please rebase onto master, and also add the Acked-by: tags from Etienne and me?

File: libteec/include/tee_client_api.h [hunk 2]
```diff
@@ -384,8 +388,10 @@ typedef union {
  */
 typedef struct {
 	/* Implementation defined */
-	TEEC_Context *ctx;
-	uint32_t session_id;
+	struct{
```
Comment (jenswi-linaro) [inline libteec/include/tee_client_api.h:391]: Please add a space (` `), between `struct` and `{`.

**Comment by jforissier** [comment]
Ping @JLLinaro

File: libteec/include/tee_client_api.h [hunk 3]
```diff
@@ -293,16 +295,18 @@ typedef struct {
...
-	} internal;
+	struct {
+		int id;
+		size_t alloced_size;
+		void *shadow_buffer;
+		int registered_fd;
+		union {
+			bool dummy;
+			uint8_t flags;
+		} internal;
```
Comment (vincent-mailhol) [inline libteec/include/tee_client_api.h:308]: Is the `dummy` field still needed following this change? `dummy` was introduced in 4f3d4cbb7824bab8f5c0a445f301528a53142f6b not to break the ABI. Here we are breaking it anyway, so while at it, I suggest some clean up:
```suggestion
		uint8_t flags;
```
*N.B.:* If there is a wish to keep more flag bits for future use, then just declare `flags` as `uint32_t`.

Comment (jforissier) [reply libteec/include/tee_client_api.h:308]: I agree. `uint32_t flags;` LGTM.

**Comment by jforissier** [comment]
@JLLinaro any chance this could make it into the next release? (-rc1 is in 3 days).

On the other hand, I am now wondering if this is a good idea. Who cares about this? Sure it is in the spec, but what difference does it make? Apps are not supposed to manipulate the `imp` field anyways. And changing it now will have impacts on downstream users (version change).

**Comment by jenswi-linaro** [comment]
The answer to your question is in https://github.com/OP-TEE/optee_client/issues/348

**Comment by jforissier** [comment]
> The answer to your question is in #348

Ah, yes. It would be nice to mention that the `imp` struct makes it easier to create a biding for Rust.

**Comment by jenswi-linaro** [comment]
I think we need to explain to @JLLinaro that it's going to cost him a beer each time github reminds us that it's not done yet. :-)

File: libteec/include/tee_client_api.h (hunk 2 above)
Comment (jforissier) [reply libteec/include/tee_client_api.h:391]: @JLLinaro comment not addressed

File: libteec/include/tee_client_api.h (hunk 3 above)
Comment (jforissier) [reply libteec/include/tee_client_api.h:308]: @JLLinaro comment not addressed

File: libteec/include/tee_client_api.h [hunk 4]
```diff
@@ -293,16 +295,17 @@ typedef struct {
...
-		uint8_t flags;
-	} internal;
+	struct {
+		int id;
+		size_t alloced_size;
+		void *shadow_buffer;
+		int registered_fd;
+		union {
+			uint32_t flags;
+		} internal;
```
Comment (jforissier) [inline libteec/include/tee_client_api.h:307]: The union should not be needed.

**Comment by jforissier** [comment]
Merged manually (with commit message slightly edited).

**Pull request closed without merging**